### PR TITLE
Treat transient connections as opt-in when opening new streams

### PR DESCRIPTION
**Pull request by vyzo**
This adds support for transient connections with opt-in semantics; when the best connection to a peer is transient, then only open a stream if the user is prepared to deal with it and has specified the relevant context option.

Rationale: many protocols may not be prepared to deal with transient connections; this makes it explicit opt-in to protect them and also eliminate interference with connection upgrade attempts.

depends on https://github.com/libp2p/go-libp2p-core/pull/175

### TODO
- [x] Release & update deps.

### CODE REVIEW

**Comment by aarshkshah1992** [comment]
@vyzo 

We still need the code to mark limited Relay connections as Transient. How and where in the code do you plan to do that without leaking the idea of limited Relays to the Swarm ?

**Comment by vyzo** [comment]
The relay dialer/listener constructs the Conn object, so it can mark itself when it knows it's transient.
No need to leak into the swarm code at all!

**Comment by vyzo** [comment]
retract that, the swarm creates the stat object -- we need a mechanism to pass stuff from the transport all the way up to the swarm.

**Comment by vyzo** [comment]
added threading of stat if available from the transport conn.

File: swarm.go [hunk 1]
```diff
@@ -351,6 +358,13 @@ func (s *Swarm) NewStream(ctx context.Context, p peer.ID) (network.Stream, error
 				return nil, err
 			}
 		}
+
+		if c.Stat().Transient {
```
Comment (aarshkshah1992) [inline swarm.go:362]: can you add a unit test for this ?

**Review by aarshkshah1992 — COMMENTED** [review]
Lgtm. Just needs a unit test.

**Review by aarshkshah1992 — APPROVED** [review]

File: swarm.go (hunk 1 above)
Comment (vyzo) [reply swarm.go:362]: it would be nice, but this is quite hard to write and probably not worth the effort.

**Comment by aarshkshah1992** [comment]
@vyzo 
 
Should I go ahead and merge this ?

**Comment by vyzo** [comment]
waiting for travis, I will merge and cut release.

**Pull request merged**